p2p/libp2p/discovery/optimizedKadDhtDiscoverer.go

```diff
@@ -0,0 +1,231 @@
+package discovery
+
+import (
+	"context"
+	"time"
+
+	"github.com/ElrondNetwork/elrond-go/p2p"
+	"github.com/libp2p/go-libp2p-core/protocol"
+	dht "github.com/libp2p/go-libp2p-kad-dht"
+)
+
+type discovererStatus string
+
+const statNotInitialized discovererStatus = "not initialized"
+const statInitialized discovererStatus = "initialized"
+const minIntervalForSeedersReconnection = time.Second
+const optimizedKadDhtName = "optimized kad-dht discovery"
+
+type optimizedKadDhtDiscoverer struct {
+	kadDHT                      KadDhtHandler
+	peersRefreshInterval        time.Duration
+	seedersReconnectionInterval time.Duration
+	protocolID                  string
+	initialPeersList            []string
+	bucketSize                  uint32
+	routingTableRefresh         time.Duration
+	hostConnManagement          *hostWithConnectionManagement
+	sharder                     Sharder
+	status                      discovererStatus
+	chanInit                    chan struct{}
+	errChanInit                 chan error
+	chanConnectToSeeders        chan struct{}
+	chanDoneConnectToSeeders    chan struct{}
+	initKadDhtHandler           func(ctx context.Context) (KadDhtHandler, error)
+}
+
+// NewOptimizedKadDhtDiscoverer creates an optimized kad-dht discovery type implementation
+// initialPeersList can be nil or empty, no initial connection will be attempted, a warning message will appear
+// This implementation uses a single process loop function able to carry multiple tasks synchronously
+func NewOptimizedKadDhtDiscoverer(arg ArgKadDht) (*optimizedKadDhtDiscoverer, error) {
+	sharder, err := prepareArguments(arg)
+	if err != nil {
+		return nil, err
+	}
+
+	if arg.SeedersReconnectionInterval < minIntervalForSeedersReconnection {
+		return nil, p2p.ErrInvalidSeedersReconnectionInterval
+	}
+
+	okdd := &optimizedKadDhtDiscoverer{
+		sharder:                     sharder,
+		peersRefreshInterval:        arg.PeersRefreshInterval,
+		seedersReconnectionInterval: arg.SeedersReconnectionInterval,
+		protocolID:                  arg.ProtocolID,
+		initialPeersList:            arg.InitialPeersList,
+		bucketSize:                  arg.BucketSize,
+		routingTableRefresh:         arg.RoutingTableRefresh,
+		status:                      statNotInitialized,
+		chanInit:                    make(chan struct{}),
+		errChanInit:                 make(chan error),
+		chanConnectToSeeders:        make(chan struct{}),
+		chanDoneConnectToSeeders:    make(chan struct{}),
+	}
+
+	okdd.initKadDhtHandler = okdd.initKadDht
```

Comment:
Why do it like this? Do we plan to allow to inject this handler from other places? Otherwise I don't see the point of a function pointer here.

Reply:
Was impossible to test otherwise